pkg/cmd/server/bootstrap/get_job_runner.go

```diff
@@ -0,0 +1,446 @@
+		WorkspaceStore:         workspaceStore,
+		WorkspaceMetadataStore: workspaceMetadataStore,
+		FindTarget: func(ctx context.Context, targetId string) (*models.Target, error) {
+			t, err := targetService.GetTarget(ctx, &stores.TargetFilter{IdOrName: &targetId}, false)
+			if err != nil {
+				return nil, err
+			}
+			return &t.Target, nil
+		},
+		FindContainerRegistry: func(ctx context.Context, image string) (*models.ContainerRegistry, error) {
+			return containerRegistryService.FindByImageName(image)
+		},
+		FindCachedBuild: func(ctx context.Context, w *models.Workspace) (*models.CachedBuild, error) {
+			validStates := &[]models.BuildState{
+				models.BuildStatePublished,
+			}
+
+			build, err := buildService.Find(&stores.BuildFilter{
+				States:        validStates,
+				RepositoryUrl: &w.Repository.Url,
+				Branch:        &w.Repository.Branch,
+				EnvVars:       &w.EnvVars,
+				BuildConfig:   w.BuildConfig,
+				GetNewest:     util.Pointer(true),
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			if build.Image == nil || build.User == nil {
+				return nil, errors.New("cached build is missing image or user")
+			}
+
+			return &models.CachedBuild{
+				User:  *build.User,
+				Image: *build.Image,
+			}, nil
+		},
+		GenerateApiKey: func(ctx context.Context, name string) (string, error) {
+			return apiKeyService.Generate(models.ApiKeyTypeWorkspace, name)
+		},
+		RevokeApiKey: func(ctx context.Context, name string) error {
+			return apiKeyService.Revoke(name)
+		},
+		ListGitProviderConfigs: func(ctx context.Context, repoUrl string) ([]*models.GitProviderConfig, error) {
+			return gitProviderService.ListConfigsForUrl(repoUrl)
+		},
+		FindGitProviderConfig: func(ctx context.Context, id string) (*models.GitProviderConfig, error) {
+			return gitProviderService.GetConfig(id)
+		},
+		GetLastCommitSha: func(ctx context.Context, repo *gitprovider.GitRepository) (string, error) {
+			return gitProviderService.GetLastCommitSha(repo)
+		},
+		CreateJob: func(ctx context.Context, workspaceId string, action models.JobAction) error {
+			return jobService.Save(&models.Job{
+				ResourceId:   workspaceId,
+				ResourceType: models.ResourceTypeWorkspace,
+				Action:       action,
+				State:        models.JobStatePending,
+			})
+		},
+		TrackTelemetryEvent:   telemetryService.TrackServerEvent,
+		ServerApiUrl:          util.GetFrpcApiUrl(c.Frps.Protocol, c.Id, c.Frps.Domain),
+		ServerVersion:         version,
+		ServerUrl:             headscaleUrl,
+		DefaultWorkspaceImage: c.DefaultWorkspaceImage,
+		DefaultWorkspaceUser:  c.DefaultWorkspaceUser,
+		Provisioner:           provisioner,
+		LoggerFactory:         loggerFactory,
+	})
+
+	return workspace_runner.NewWorkspaceJobRunner(workspace_runner.WorkspaceJobRunnerConfig{
+		FindWorkspace: func(ctx context.Context, workspaceId string) (*models.Workspace, error) {
+			workspaceDto, err := workspaceService.GetWorkspace(ctx, workspaceId, false)
```

Comment:
I think it makes more sense to call `server.GetInstance(nil).WorkspaceService...` (and for all other services) than to construct them here all over again.